### PR TITLE
Machine reassignment constraint streams

### PR DESCRIPTION
**Pull request by rsynek**

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MrConstraintName.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.machinereassignment.solver.score;
+
+enum MrConstraintName {
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MrConstraintName.java:19]: Intresting that you're using an enum. It crossed my mind too.

For other examples, we've always just use public static strings - it should be consistent (either we refactor them too or we refactor this).
The problem with enums is that the Solver API expects a String constraintName, not an enum - so you'd need to translate every string coming out of the Indictement/Match api first - which is overhead to just keeping it as a String. Wdyt?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    private void transientUsage(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.TRANSIENT_USAGE.getName(),
+                                                                          HardSoftLongScore.ofHard(1L));
+        constraint.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      Joiners.equalTo(MrMachineCapacity::getMachine, MrProcessAssignment::getOriginalMachine)
+                )
+                .filter(
+                        (mrMachineCapacity, mrProcessAssignment) -> mrProcessAssignment.isMoved() && mrMachineCapacity.isTransientlyConsumed()
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: soft suggestion: For readability, I 'd remove the \n after filter( and add one before &&

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    /**
+     * Process move cost: A process has a move cost.
+     */
+    private void processMoveCost(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.PROCESS_MOVE_COST.getName(),
+                                                                          HardSoftLongScore.ofSoft(1L));
+        constraint.from(MrProcessAssignment.class)
+                .filter(processAssignment -> processAssignment.isMoved())
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Shall we use ProcessAssignment::isMoved?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    /**
+     * Conflict: Processes of the same service must run on distinct machines.
+     */
+    private void serviceConflict(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.SERVICE_CONFLICT.getName(),
+                                                                          HardSoftLongScore.ofHard(1L));
+        constraint.from(MrProcessAssignment.class)
+                .joinOther(
+                        Joiners.equalTo(MrProcessAssignment::getMachine),
+                        Joiners.lessThan(MrProcessAssignment::getId)
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: joinOther() implies the lessThan(MrProcessAssignment::getId) automatically, see javadoc

Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Ok, that javadoc doesn't really clearly mention that. I improving that now :)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MrConstraintName.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MrConstraintName.java:19]: Is there a good reason for an interface with constants these days? It comes from the time before enums, AFAIK.
There is no problem with keeping the Solver API as is; one just needs to pass there the enum's string representation. Maybe I don't fully understand the issue.
As to the consistency, I am definitely not turning all examples to enums as a part of this PR to your fork of OptaPlanner :) let's do it in later in a PR that goes to optaplanner/master.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java (hunk 4 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Interesting, so I need to provide just the first joiner.

Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: javadoc improved on my branch, rebase on to confirm the javadoc is now clearer :)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    private void serviceMoveCost(ConstraintFactory constraintFactory) {
+        throw new UnsupportedOperationException("Not yet implemented due to missing aggregation function.");
+
+        /*
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.SERVICE_MOVE_COST.getName(),
+                                                                          HardSoftLongScore.ofSoft(1L));
+        constraint.from(MrProcessAssignment.class)
+                .filter(processAssignment -> processAssignment.isMoved())
+                .groupBy(processAssignment -> processAssignment.getService(), ConstraintCollectors.count())
+                .penalizeLong();
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java:203]: This is missing a max or something, right? Let's add a TODO there so it sticks out

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MrConstraintName.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MrConstraintName.java:19]: Do you insist on turning this into an interface with constants and we rewrite everything later? For me, having an enum here and constants everywhere else is not a big issue, especially if that's just a temporary state.

**Pull request merged**